### ISSUE TITLE
isset(WC_Shipping_Rate->meta_data['some-custom-meta']) always returns false

### ISSUE THREAD

**Issue by shakethatweight-simon** [woocommerce/woocommerce#28072]
**Prerequisites (mark completed items with an [x]):**
- [x] I have have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
`isset()` fails when checking meta_data in Shipping rates due to the `__isset()` method only checking the data property of the object
https://github.com/woocommerce/woocommerce/blob/3f61afe5337c4fbb94a6b14b919cda3795ea4e05/includes/class-wc-shipping-rate.php#L67-L69

**Expected behaviour**
`isset()` should return true/false depending on the existence of the value within the meta_data property

**Actual behaviour**
`isset()` always returns false, whether the meta_data value has been set or not, due to `WC_Shipping_Rate::__isset` only checking if `meta_data` `isset` within the `data` property, but that's not where it's saved.

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:

```
// $this is a class which has extended from WC_Shipping_Method for a custom shipping rates plugin
$this->add_rate([
    'id' => 'my-rate',
    'label' => 'my label',
    'cost' => 1.5,
    'meta_data' => [
        'custom-meta-1' => 'some-value'
    ]
]);
```

Later code, for example in the cart-shipping.php template.

```
isset($method->meta_data['custom-meta-1']) : always returns false for any meta_data field
$method->meta_data['custom-meta-1'] : does correctly return the value if it's been set
```
https://github.com/woocommerce/woocommerce/blob/3f61afe5337c4fbb94a6b14b919cda3795ea4e05/includes/class-wc-shipping-rate.php#L78-L86

`$method->meta_data['custom-meta-1']` does work, due to `__get` first checking for the existence of the method get_meta_data.
If `__isset` is updated to also check to see if `get_$key` method exists, and return true, php would then do a final check on isset($method->get_meta_data()['custom-meta-1']) for the final test to see if the requested value is set

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

This section of code in the WC_Shipping_Rates class has been untouched in 3 years, surprised nobody else has stumbled across this problem before now.

Suggestion: 

```
	public function __isset( $key ) {
		if ( is_callable( array( $this, "get_{$key}" ) ) ) {
			return true;
		}
		return isset( $this->data[ $key ] );
	}
```

**Comment by peterfabian** [comment]
Thanks for the bug report, @shakethatweight-simon ! 

I agree there is an inconsistency, but I'm not sure if the intention behind the code was to make the `meta_data` available in this way via magic methods. Otherwise, there would probably be a method called `set_meta_data` that would allow setting the meta_data, similarly to how the `get_meta_data` allows getting it. 

I think in this case, it might be more of a coincidence/oversight that it works. I quickly scanned our codebase and couldn't find any usage where we would directly use `$shipping_rate->meta_data`, but maybe I missed something. 

Since we can't really rename public methods, I think we have 2 options:
- 1. add `set_meta_data` method and modify the `__isset` magic method as you described, making `meta_data` a regularly accessible property (this also adds the possibility to replace meta_data, which wasn't possible before).
- 2. exclude `get_meta_data` from `__get` and make it 'more private', only accessible via `get_meta_data` and `add_meta_data` methods.

No 1 goes slightly against our principles of reducing exposure of too much internals via our API/classes (it's almost as if we made $meta_data public), and no 2 is slightly backward incompatible. 

I'm not totally sure either of those is a good tradeoff vs having a slightly inconsistent class `WC_Shipping_Rate` interface.

**Comment by shakethatweight-simon** [comment]
Thanks for the detailed response @peterfabian, I can understand your thoughts on this. Could we at least update the `__isset` method to use `_doing_it_wrong` when `$key == 'meta_data'` ? No problems with any backwards compatibility issue, and any developer working with this in future would at least notice the warning when trying to use it the wrong way?

**Comment by shakethatweight-simon** [comment]
> do you see this being used by WC core or any other plugins
No, I noticed it when I was updating a custom shipping plugin and couldn't work out why isset and empty where giving me unexpected results.

**Comment by peterfabian** [comment]
> Could we at least update the __isset method to use _doing_it_wrong when $key == 'meta_data' ? No problems with any backwards compatibility issue, and any developer working with this in future would at least notice the warning when trying to use it the wrong way?

That seems like a good solution, thanks. Would you like to open a PR since you reported it and came up with it? If not, we can do it no problem.

**Comment by shakethatweight-simon** [comment]
> That seems like a good solution, thanks. Would you like to open a PR since you reported it and came up with it? If not, we can do it no problem.

Just noticed there is a `wc_doing_it_wrong` I'll use that one instead of the Wordpress equivalent.

**Comment by shakethatweight-simon** [comment]
@peterfabian or not... just noticed that it's listed in `includes/wc-deprecated-functions.php` and most of the code currently uses `_doing_it_wrong` instead of `wc_doing_it_wrong`. Not sure of the history of this function, which one do you want me to use?

**Comment by peterfabian** [comment]
I think `wc_doing_it_wrong` is just a utility function that was placed in a file that houses 
- utility functions used for deprecations, and 
- deprecated functions themselves. 

It doesn't have tag `@deprecated`, so I think it's not actually deprecated, so feel free to use `wc_doing_it_wrong`.

**Comment by fean** [comment]
Hey guys!

I think we're also experiencing this same issue.

Our `WC_Shipping_Method` derived class we set metadata in the `calculate_shipping` method override like so:
```php
public function calculate_shipping($package = [])
{
   $rate = [
        'id' => $this->id,
        'label' => $this->title,
        'cost' => 7,
        'meta_data' => [
            'cutOffTime' => '11:00',
        ],
    ];

    $this->add_rate($rate);
}
```

Later on in the filter `woocommerce_cart_shipping_method_full_label` when we want to render a label with the `meta_data`, we are not able to access it as the `WC_Shipping_Rate` instance has no entries in `meta_data`.

Assuming we have a variable `$rate` which holds an the instance of `WC_Shipping_rate`.
We expect `$rate->meta_data['cutOffTime']` to yield `11:00`.

Are we doing something wrong or is this the same bug as described by @shakethatweight-simon? Thanks for your help guys!

**Comment by shakethatweight-simon** [comment]
In my experience, the meta value was there, but isset incorrectly reported that it didn't exist. So I was still able to use `$rate->meta_data['cutOffTime']` - but couldn't use `isset($rate->meta_data['cutOffTime'])` 

I've not touched this section of code since the original report, so not sure if something has changed since then.

**Comment by fean** [comment]
You're absolutely right! I'm very new to WooCommerce development and didn't know the shipping rate is cached. The `calculate_shipping` method is only changed when the cart details or address details have been updated.

I'm sorry for the inconvenience! 🙂

**Comment by ghost** [comment]
So, how does one check if the meta_key is set? any workaround?

        add_action( 'woocommerce_after_shipping_rate', 'action_after_shipping_rate', 20, 2 );
        function action_after_shipping_rate ( $method, $index ) {
          if ( isset( $method->meta_data['meta_value'] ) ) {
           $meta_value = $method->meta_data['meta_value'];
           echo "Your meta value is: " . $meta_value;
         }
        }

The `isset()`  will always return false. Even if `meta_data['meta_value']` is in fact set.

As already mentioned above, `$meta_value` will correctly hold `$method->meta_data['meta_value'];` value if there is one. Issue is it will, understandably. throw an error if the value is not set. We need another way to check if the value is set first but can't seem to think of any.

Any workaround to check for if  meta_key is set ?

Update: Apparently, using `array_key_exists()` instead of `isset()` works.

**Comment by shakethatweight-simon** [comment]
> So, how does one check if the meta_key is set? any workaround?

...

> Update: Apparently, using `array_key_exists()` instead of `isset()` works.

You could also try `isset($method->get_meta_data()['meta_value'])` - not tested, but from memory this should work.
